runners/flink/src/main/java/org/apache/beam/runners/flink/adapter/BeamAdapterUtils.java

```diff
@@ -76,7 +76,8 @@ Map<String, DataSetOrStreamT> applyBeamPTransformInternal(
                         new FlinkInput<>(
                             key,
                             BeamAdapterCoderUtils.typeInformationToCoder(
-                                getTypeInformation.apply(Preconditions.checkNotNull(flinkInput)), coderRegistry)))));
+                                getTypeInformation.apply(Preconditions.checkNotNull(flinkInput)),
```

Comment:
how is it that flinkInput could be nullable here anyhow?

Reply:
So the crux of the problem (and why I went with suppressing warnings in the first place, here and elsewhere) is the delta between the nullness checker flagging an issue and there actually being an issue. 

The inferred type of flinkInput is `@Nullable` because a function that accepts nullable arguments is perfectly to use when transforming maps that are known to not have nullable types. But the nullchecker simply isn't smart enough to figure that out, hence the need to inform it directly.

Reply:
Yes, that crux is just an argument about types in general, not specific to null checks. In modern typing, `@Nullable Foo` and `Foo` are different types, and you should know which one you are dealing with.

FWIW the nullness checker is simply a fully rigorous type system, and significantly smarter than, say, Java generic inference. It can definitely infer polymorphism of nullability (I haven't dug into this particular line of code). The problem you are describing sounds like the fact that Java does not implicitly extend type variables, so if you have, say, a `Collection<Foo>` and you want to map over it, the `map` function has to accept `Function<X super Foo, ...>` and otherwise won't be able to accept variations.

But either way, I know this is hard in general, especially interacting with legacy codebases. So putting targeted suppressions with explanations is OK, though adding checks so that the error is more localized is better.

Here is the biggest reason I push on suppression of type checking: it will suppress it for the other lines that are not part of what you are trying to deal with, and it will also suppress it for future code. And of course when someone comes across the code they will think they should just go ahead and suppress it whenever they want.

So ideally suppress would be on just one single line that adjusts types from what is inferred to what you want them to be. But you can usually do that with a check.

Reply:
lol I think I got the direction of variance wrong in the `Function<X extends Foo, ...>`. It is a pain and easy to get wrong, which is why we have type checkers of course :-)

My other take is that if the type checker cannot understand it, a person coming to the code with no context also probably cannot understand it easily. I'm sure _you_ understand it well enough, especially at the time of authoring the code :-)

Reply:
I think humans and typecheckers are (fortunately) somewhat complimentary in their validation capabilities. It took a while to figure out why the typechecker couldn't figure things out here, and it boils down to "it's not smart enough/the language isn't rich enough to express the full intent here." 

I also think that part of the problem with all this code is that Java decided to use nullness to indicate the lack of an item in a Map rather than throw an error, and the type system isn't close to sophisticated to identify maps that have certain keys. (One could argue that that's a reason to avoid Maps. Yes, it is, but this is solving a generic case with specializations that hide the warning suppressions to use the less safe API.)

I agree with trying to be narrow with what is suppressed, which is why I didn't just slap it on the entire class. (Well, I did when developing, 'cause it was just to painful otherwise, but once I'd cleaned it up.)

Reply:
I mean, checkerframework does have a type system for tracking keys that exist in maps but you don't want to go there :-)

The basic truth is that you generally should always check map return results for null. Even if today there is a clear invariant that it will be in there, that will get broken by people down the line. A message saying what is corrupted is great.